Python 3 compatibility updates

drsteve opened the pull request:
Closes #25 

- Make all relative imports explicit rather than implicit
- Fixed all statement vs function issues (e.g. `print`)
- Moved tests to `test` directory

Note that in a conda env the install path will need to be added to PYTHONPATH, as the Python wrappers are not currently installed using a `setup.py`.

drsteve commented:
Currently seeing a number of errors that may be (???) associated with incompatibilities between ctypesgen and Python 3.7.

Main error messages look like:
```
  File "[...]/local/lib/python3.7/site-packages/lgmpy/Lstar.py", line 278, in get_Lstar
    mmi = MagEphemInfo.LstarInfo.contents.mInfo.contents
ValueError: NULL pointer access
```

I also see a failure that I haven't dug into:
```
======================================================================
FAIL: test_size_Lgm_MagModelInfo (test_Lgm_MagModelInfo.Lgm_MagModelInfoTests)
for Lgm_MagModelInfo the c and python sizes should match
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/net/toaster/u/smorley/projects/LANLGeoMag/Python/test/test_Lgm_MagModelInfo.py", line 30, in test_size_Lgm_MagModelInfo
    ctypes.sizeof(Lgm_MagModelInfo.Lgm_MagModelInfo) )
AssertionError: 872288 != 872352
```
@balarsen @jtniehof, any thoughts?

jtniehof commented:
Those are almost certainly related. Probably the struct padding convention has changed at some point. Which one of those is "expected" and which "actual"?

Make sure everything is really rebuilt from scratch (clear out all the generated wrappers and all the C), and try a pull of a new ctypesgen as well. If that doesn't fix it we're in for heavy sailing.

jtniehof commented:
Note I have ctypesgen update info in HACKING. I'm not sure if https://github.com/davidjamesca/ctypesgen/ or https://github.com/olsonse/ctypesgen/ is better. The latter is what we were using and is a fork of the former, but the former does seem to be updated.

drsteve commented:
To come back to this... (user inquiries floated it back onto my list)
We were using olsonse's fork because it was being maintained and appeared to be the _de facto_ ctypesgen. It looks like olsonse has been able to run the original fork, so everything is level again. 

Things seem to be working under Python 3.6, but I'm not sure that ctypesgen plays nicely with 3.7 or 3.8 yet. I'll try setting up some new virtual environments and build under each... 
We do, however, now have some Python test failures (under 3.6) that will need further investigation. My first guess is that they're all due to the recent update of the IGRF (and the bugfix in one coefficient).

drsteve commented:
I rebased against master to make sure all the IGRF updates were properly reflected, and I've updated the tests that were impacted by that change.

The Python tests all pass now on 3.7.5. Since Python 2 is past EOL, I figured I should open this up for merge and then we can tag a new release.

jtniehof commented:
Sorry, somehow this flew right off my radar (and I was checking back here for other reasons.) I'll dig in next week.

jtniehof commented on Python/lgmpy/Lgm_Vector.py:
```diff
@@ -241,13 +246,39 @@ def __div__(self, other):
+    __truediv__ = __div__
+
+    def __floordiv__(self, other):
+        """
+        divide a scalar into each component
+
+        @param other: scalar to divide into vector
+        @type other: (int, long, float)
+
+        @return: new vector
+        @rtype: Lgm_Vector
+
+        @author: Brian Larsen
+        @organization: LANL
+        @contact: balarsen@lanl.gov
+
+        @version: V1: 22-Dec-2010 (BAL)
+        """
```
(I had a crash so this might be a duplicate comment on this, in which case the other one is better.)

The docstring should probably be updated (for author and the fact that it's floor div). What's the method resolution order on Py2k and Py3k? Will 2k always do `div` for `/` and `floordiv` for `//`, and 3k do `truediv` for `/` and `floordiv` for `//`? Just wondering about the assignment and if it's the right way around. I'm also going to suggest a future import for division elsewhere; that might affect this.

drsteve replied:
See previous. Marking as resolved since the comment is a duplicate.

jtniehof commented on Python/test/test_Closed_Field.py:
```diff
@@ -52,14 +52,15 @@ def test_extended_out(self):
         self.assertEqual(ans, 'LGM_CLOSED')
         numpy.testing.assert_allclose(north_fp,
-                [-0.0620421299, 0.39455411901, 0.9306480813], rtol=1e-5)
+                [-0.062047, 0.394557, 0.930652], rtol=1e-5)
```
I presume this and other number changes are based on the new IGRF?

drsteve replied:
Yes.

jtniehof commented on Python/test/test_Lgm_Vector.py:
```diff
@@ -274,7 +274,9 @@ def test_SphToCart(self):
         self.assertRaises(ValueError, Lgm_Vector.SphToCart, [1]*2, [2]*3, [3]*2)
         # test putting in lists
-        ans_tst = Lgm_Vector.SphToCart(zip(*invals)[0], zip(*invals)[1], zip(*invals)[2] )
+        ans_tst = Lgm_Vector.SphToCart(list(zip(*invals))[0],
+                                       list(zip(*invals))[1],
+                                       list(zip(*invals))[2])
```
This is hardly optimization of the century, but wouldn't it make sense to do list(zip(*invals)) once, store the value, and then index it three times?

drsteve replied:
It would, although this is just a formatting change over the previous incarnation - I didn't change code here.
I can pull the zip out onto a previous line and then index the list in the call.

jtniehof replied:
Whichever way you think is easier to read, honestly.

jtniehof commented on Python/test/test_magcoords.py:
```diff
@@ -51,8 +50,8 @@ def test_end2end_jpl_eph(self):
-                 "2015-03-11T11:58:40.170326 479347187354325578 GSE2000 SM -20164.747998 -52414.140552 -20241.497772 -20870.283837 -46078.983969 -31697.569507",
-                 "2015-03-12T04:25:43.605928 479406410789928239 SM GSM 75645.819112 -40422.353445 -35182.256254 81628.221639 -40422.353445 -17231.208208"]
+                 "2015-03-11T11:58:40.170326 479347187354325578 GSE2000 SM -20164.747998 -52414.140552 -20241.497772 -20868.064112 -46078.015440 -31700.438751",
+                 "2015-03-12T04:25:43.605928 479406410789928239 SM GSM 75645.819112 -40422.353445 -35182.256254 81628.244132 -40422.353445 -17231.101651"]
```
Should the C tests also be updated (or are they already?)

drsteve replied:
C tests were already updated.

jtniehof reviewed (commented):
Does it make sense to do future imports for relative import, division, and print function? That will make the Python 2 work more like the eventual Python 3.

Overall this is looking good, see the comments for suggestions.

drsteve commented on Python/lgmpy/Lgm_CTrans.py:
```diff
@@ -156,9 +156,8 @@ def dateToFPHours(inval):
-        lst = [val.hour + val.minute/60 +
-                                val.second/60/60 +
-                                val.microsecond/60/60/1000000 for val in inval]
+        lst = [val.hour + val.minute/60 + val.second/60/60 +
+               val.microsecond/60/60/1000000 for val in inval]
```
Yes. Definitely should be float division. There's a `from __future__ import division` at the top of `Lgm_Ctrans.py` so behaviour shouldn't have changed. This is also covered by unit testing.

jtniehof replied:
Got it; thanks. Should have looked at more context.

drsteve commented on Python/lgmpy/Lgm_CTrans.py:
```diff
@@ -80,12 +80,12 @@ def dateToDateLong(inval):
     -------
-    out : long, list
-        long or list of longs corresponding to the datetime
+    out : int, list
+        int or list of ints corresponding to the datetime
```
I believe your logical chain is right. It's a C long, so the translation was to Python long, which went away for Python 3. You raise a good point though - we should check whether this change actually passes everything through to LGM correctly under Python 2.
The docstring also has a suffixed `L`  so I'll need to update that.

drsteve replied:
@jtniehof : I was checking on this and was thinking about a "better safe than sorry", most likely overkill, change to ensure that 2.7 is still supported.

Basically, where the code here changed from `long(...)` to `int(...)` it's likely that this isn't Python 2.7 compatible (but I don't have a good test environment right now).
However, if `long` isn't defined then it raises a NameError. Any objection to:
```
def longint(inval):
    try:
        return long(inval)
    except NameError:
        return int(inval)
```
Then replace the `long` call with a call to `longint`?

jtniehof replied:
I don't think you need to go to that length. I think for `dateToDateLong`, a Python `int` is appropriate. Python `int` and `long` aren't the same as C `int` and `long` and `ctypes` will make sure they're converted to the appropriate C ctype on a call. So the only remaining thing is whether the values can fit in Python 2 ints. The maximum reasonable value seems to be 99991231, which fits in a 32-bit signed number just fine. Should be no need for a Python 2 Long then; I don't think we're running on 16-bit machines.

Short is I think nothing else needs to be done beyond your initial commit. I'm not going to mark this as resolved, but you should if you agree.

drsteve commented on Python/lgmpy/Lgm_Vector.py:
```diff
@@ -11,20 +11,25 @@
 __author__ = 'Brian Larsen (Python), Mike Henderson (C) - LANL'
 
 class Lgm_Vector(Lgm_Vector):
+    if sys.version_info.major<3:
+        _typetup = (int, float, long)
+    else:
+        _typetup = (int, float)
```
This seemed like the easy way to do it when I was fixing, as the code already checked that type was one of the members of a tuple. This change just set the tuple at the top of the class and drops `long` under Python 3.
if you have a cleaner suggestion then I'm happy to implement - this was the minimal fix without refactoring the class.

jtniehof replied:
My thought was:
```python
try:
    _typetup = (int, float, long)
except NameError: #Python 3
    _typetup = (int, float)
```
It obscures things if there's a _different_ `NameError`, of course....

drsteve replied:
Leaving this _as-is_ for now. The python wrappers need a major overhaul and I doubt this will survive that. It currently works as intended, so marking as resolved.

drsteve commented on Python/lgmpy/Lgm_Vector.py:
```diff
@@ -241,13 +246,39 @@ def __div__(self, other):
+    __truediv__ = __div__
+
+    def __floordiv__(self, other):
+        """
+        divide a scalar into each component
+
+        @param other: scalar to divide into vector
+        @type other: (int, long, float)
+
+        @return: new vector
+        @rtype: Lgm_Vector
+
+        @author: Brian Larsen
+        @organization: LANL
+        @contact: balarsen@lanl.gov
+
+        @version: V1: 22-Dec-2010 (BAL)
+        """
```
Python 3: `/` calls `__truediv__` and `//` calls `__floordiv__`
Python 2: `//` calls `__floordiv__`, and how `/` resolves depends on whether future division is imported.

This file has `from __future import division` at the top already, so behaviour should be consistent across 2.7 and 3.x..

jtniehof replied:
I think all is well then, and should just update docstring here to be you as author and note that the division is floor (yes, the name implies that, but still....)
